backend/LexBoxApi/GraphQL/ProjectMutations.cs

```diff
@@ -100,53 +101,69 @@ public record CreateProjectResponse(Guid? Id, CreateProjectResult Result);
     }
 
     public record UserProjectRole(string Username, ProjectRole Role);
-    public record BulkAddProjectMembersResult(List<UserProjectRole> AddedMembers, List<UserProjectRole> CreatedMembers, List<UserProjectRole> ExistingMembers);
+    public record BulkAddProjectMembersResult(List<UserProjectRole> AddedMembers, List<UserProjectRole> CreatedMembers, List<UserProjectRole> ExistingMembers, List<UserProjectRole> InvitedMembers);
 
     [Error<NotFoundException>]
+    [Error<FormatException>]
     [Error<DbError>]
     [AdminRequired]
     [UseMutationConvention]
     public async Task<BulkAddProjectMembersResult> BulkAddProjectMembers(
         LoggedInContext loggedInContext,
         BulkAddProjectMembersInput input,
-        LexBoxDbContext dbContext)
+        LexBoxDbContext dbContext,
+        [Service] EmailService emailService)
     {
         var project = await dbContext.Projects.FindAsync(input.ProjectId);
         if (project is null) throw new NotFoundException("Project not found");
         List<UserProjectRole> AddedMembers = [];
         List<UserProjectRole> CreatedMembers = [];
         List<UserProjectRole> ExistingMembers = [];
+        List<UserProjectRole> InvitedMembers = [];
         var existingUsers = await dbContext.Users.Include(u => u.Projects).Where(u => input.Usernames.Contains(u.Username) || input.Usernames.Contains(u.Email)).ToArrayAsync();
         var byUsername = existingUsers.Where(u => u.Username is not null).ToDictionary(u => u.Username!);
         var byEmail = existingUsers.Where(u => u.Email is not null).ToDictionary(u => u.Email!);
-        foreach (var username in input.Usernames)
+        foreach (var usernameOrEmail in input.Usernames)
         {
-            var user = byUsername.GetValueOrDefault(username) ?? byEmail.GetValueOrDefault(username);
+            var user = byUsername.GetValueOrDefault(usernameOrEmail) ?? byEmail.GetValueOrDefault(usernameOrEmail);
             if (user is null)
             {
                 var salt = Convert.ToHexString(RandomNumberGenerator.GetBytes(SHA1.HashSizeInBytes));
-                var isEmailAddress = username.Contains('@');
-                // TODO: In the future we'll want to allow usernames in the form "Real Name <email@example.com>" and extract the real name from them
-                // For now, just:
-                var name = username;
-                user = new User
+                var isEmailAddress = usernameOrEmail.Contains('@');
+                if (isEmailAddress)
+                {
+                    string? email;
+                    try {
+                        var parsed = new MailAddress(usernameOrEmail);
+                        email = parsed.Address;
+                    } catch (FormatException) {
+                        // FormatException message from .NET talks about mail headers, which is confusing here
+                        throw new FormatException($"Invalid email address: {usernameOrEmail}");
```

Comment:
How about we make the `usernameOrEmail` a property of the exception. Then we can show a clean translated error message (like the one you have here) on the UI.

Reply:
Tried. Can't get the `extensions` property on GraphQL errors to pass through our type-generation system. Here's what I tried. In GraphQlSetupKernel.cs:

```cs
.AddErrorFilter(error =>
{
    if ((error.Exception?.Data?.Count ?? 0) > 0)
    {
        return error.WithExtensions((IReadOnlyDictionary<string, object?>)error.Exception!.Data);
    }
    return error;
})
```

Then I created an InvalidEmailException that takes a message and an address and stores the address in the exception's Data property as `Data["address"]`. But the `extensions` field isn't being added to `frontend/schema.graphql` when I run `task api:generate-gql-schema`, so I can't get the frontend UI to build if I include `extensions` in the GraphQL request. So I'm not able to get the email address to be a property of the exception in the way the spec says it's supposed to be done.

I'm going to push a commit with what I tried, another commit reverting it, then switch InvalidEmailException to pass the actual email itself (and nothing else) in the `message` property. That's a little hackish but it will work, without my having to spend lots of time fighting the GraphQL system to make it do what the docs say it should do.

Reply:
Done in commit 9304a65c.

Reply:
yeah gql doesn't support dictionary type objects as you must know all the keys ahead of time. They do support discriminated unions so it's ok. What you did is fine though.